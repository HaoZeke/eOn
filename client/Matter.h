/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#pragma once
#include "Eigen.h"
#include "Parameters.h"
#include "Potential.h"
#include <memory>
#include <spdlog/sinks/basic_file_sink.h>
#include <string>

namespace eonc {
// This is a forward declaration of BondBoost to avoid a circular dependency.
class BondBoost;

// XXX: Really need this to go..
constexpr size_t MAXC{100}; // maximum number of components for functions
                            // matter2con and con2matter

/* Data describing an atomic structure. This class has been devised to handle
 * information about an atomic structure such as positions, velocities, masses,
 * etc. It also allow to associate a forcefield for the structure through a
 * pointer to function (potential()). The class can read and save data to a .con
 * file (atom2con() and con2atom()). It can also save to a .xyz file
 * (atom2xyz()).*/

class Matter {
public:
  ~Matter() = default;
  Matter(std::shared_ptr<Potential> pot)
      : potential{pot} {
    m_log = spdlog::get("combi");
  }
  Matter(const Matter &matter);                  // create a copy of matter
  const Matter &operator=(const Matter &matter); // copy the matter object
  // bool operator==(const Matter& matter); // true if differences in positions
  // are below differenceDistance bool operator!=(const Matter& matter); //
  // inverse of ==
  bool compare(const Matter &matter, bool indistinguishable = false);

  double
  distanceTo(const Matter &matter); // the distance to the given matter object
  double
  perAtomNorm(const Matter &matter) const; // the maximum distance between two
                                           // atoms in the Matter objects
  void
  setPotential(std::shared_ptr<Potential> pot); // set potential function to use
  std::shared_ptr<Potential> getPotential();    // get potential function to use
  void resize(long int nAtoms); // set or reset the number of atoms
  size_t numberOfAtoms() const; // return the number of atoms
  Matrix3S getCell() const;
  void setCell(Matrix3S newCell);
  double getPosition(long int atom, int axis)
      const; // return the position of an atom along one of the axis
  void setPosition(
      long int atom, int axis,
      double position); // set the position of atom along axis to position
  void setVelocity(
      long int atom, int axis,
      double velocity); // set the velocity of atom along axis to velocity
  AtomMatrix pbc(AtomMatrix diff) const;
  VectorType pbcV(VectorType diff) const;

  size_t getPotentialCalls() const;
  AtomMatrix getPositions() const; // return coordinates of atoms in array pos
  VectorType getPositionsV() const;
  AtomMatrix
  getPositionsFree() const; // return coordinates of free atoms in array pos
  VectorType getPositionsFreeV() const;
  void
  setPositions(const AtomMatrix pos); // update Matter with the new positions of
                                      // the free atoms given in array pos
  void setPositionsV(const VectorType pos);
  void setPositionsFree(
      const AtomMatrix pos); // update Matter with the new positions of the free
                             // atoms given in array pos
  void setPositionsFreeV(const VectorType pos);

  AtomMatrix getVelocities() const;
  void setVelocities(const AtomMatrix v);
  void setBiasForces(const AtomMatrix bf);
  void setBiasPotential(BondBoost *bondBoost);
  void setForces(const AtomMatrix f);
  AtomMatrix getAccelerations();

  AtomMatrix getForces(); // return forces applied on all atoms in array force
  AtomMatrix getBiasForces();
  VectorType getForcesV();
  AtomMatrix getForcesFree();
  VectorType getForcesFreeV();

  double getMass(long int atom) const; // return the mass of the atom specified
  void setMass(long int atom, double mass); // set the mass of an atom
  void setMasses(VectorType massesIn);      // set the mass of an atom
  size_t getAtomicNr(
      size_t atom) const; // return the atomic number of the atom specified
  void setAtomicNr(long int atom,
                   long atomicNr);         // set the atomic number of an atom
  Vector<size_t> getAtomicNrs() const;     // Get the vector of atomic numbers
  Vector<size_t> getAtomicNrsFree() const; // Get the vector of atomic numbers
  void
  setAtomicNrs(const Vector<size_t> atmnrs); // Get the vector of atomic numbers

  int getFixed(long int atom)
      const; // return true if the atom is fixed, false if it is movable
  void setFixed(long int atom,
                int isFixed); // set the atom to fixed (true) or movable (false)
  // void setPotentialEnergy(double);
  double getEnergyVariance();
  // VectorType getForceVariance();
  // double getMaxVariance();
  double getPotentialEnergy();
  double getKineticEnergy() const;
  double getMechanicalEnergy(); // return the mechanical energy (i.e. kinetic
                                // plus potential energy)

  double distance(long index1, long index2)
      const; // return the distance between two atoms in same configuration
  double pdistance(long index1, long index2, int axis) const;
  double distance(const Matter &matter, long index)
      const; // the distance between the same atom in two cofigurations

  long int
  numberOfFreeAtoms() const; // return the number of free (or movable) atoms
  long int numberOfFixedAtoms() const; // return the number of fixed atoms

  size_t
  getForceCalls() const; // return how many force calls that have been performed
  void resetForceCalls(); // zeroing the value of force calls

  double maxForce(void);

  void writeTibble(std::string filename);
  bool con2matter(std::string filename); // read con file into Matter, return
                                         // true if successful
  bool con2matter(FILE *file); // read con file and load data into Matter,
                               // return true if successful
  bool
  convel2matter(std::string filename); // read con file with both coordinates
                                       // and velocities into Matter
  bool convel2matter(FILE *file); // read con file with both coordinates and
                                  // velocities and load data into Matter
  bool
  matter2con(std::string filename,
             bool append = false); // print con file from data in Class Matter
  bool matter2con(FILE *file);     // print con file from data in Class Matter
  bool
  matter2convel(std::string filename); // print con file with both coordinates
                                       // and velocities  in Class Matter
  bool matter2convel(FILE *file); // print con file with both coordinates and
                                  // velocities from data in Class Matter
  void matter2xyz(std::string filename,
                  bool append = false); // print xyz file from data in Matter
  AtomMatrix getFree() const;
  VectorType getFreeV() const;
  VectorType getMasses() const;

private:
  std::shared_ptr<spdlog::logger> m_log;
  std::shared_ptr<Potential> potential;
  // --- These should be parsed in
  bool usePeriodicBoundaries{true};
  // Indicates if the potential energy and forces need to be recalculated
  bool recomputePotential{true};
  bool removeNetForce{true};
  // ----
  size_t forceCalls{0};

  // CON file header information, which is not used in the eon code
  char headerCon1[512];
  char headerCon2[512];
  char headerCon4[512];
  char headerCon5[512];
  char headerCon6[512];

  void computePotential();
  void initializeDataMembers(std::shared_ptr<Parameters> parameters);
  void applyPeriodicBoundary();
  void applyPeriodicBoundary(double &component, int axis);
  void applyPeriodicBoundary(AtomMatrix &diff);

  // Stuff which used to be in MatterPrivateData
  size_t nAtoms{0};
  AtomMatrix positions{MatrixType::Zero(0, 3)};
  AtomMatrix velocities{MatrixType::Zero(0, 3)};
  AtomMatrix forces{MatrixType::Zero(0, 3)};
  AtomMatrix biasForces{MatrixType::Zero(0, 3)};
  BondBoost *biasPotential{nullptr};
  VectorType masses{VectorType::Zero(0)};
  Vector<size_t> atomicNrs{Vector<size_t>::Zero(0)};
  Vector<int> isFixed{Vector<int>::Zero(
      0)}; // array of bool, false for movable atom, true for fixed
  Matrix3S cell{Matrix3S::Zero()};
  Matrix3S cellInverse{Matrix3S::Zero()};
  double energyVariance{0.0};
  double potentialEnergy{0.0};
};

} // namespace eonc
