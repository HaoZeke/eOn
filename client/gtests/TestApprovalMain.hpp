/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#define APPROVALS_CATCH2_V3
#include "ApprovalTests.hpp"

// This puts "received" and "approved" files in approval_tests/ sub-directory,
// keeping the test source directory tidy:
const auto directoryDisposer =
    ApprovalTests::Approvals::useApprovalsSubdirectory("approval_tests");
