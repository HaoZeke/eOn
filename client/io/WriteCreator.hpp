/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#pragma once
#include <magic_enum/magic_enum_all.hpp>

#include "client/io/IOBase.hpp"

namespace eonc::io {

enum class WriteType {
  UNKNOWN = -1, // error case
  NONE = 0,
  CON,
  CONVEL,
  XYZ,
  TIBBLE
};

std::unique_ptr<WriteBase> mkWriter(const toml::table &config);
} // namespace eonc::io
