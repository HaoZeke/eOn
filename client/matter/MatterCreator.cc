/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#include "client/matter/MatterCreator.hpp"
#include "client/Eigen.h"
#include "client/Element.hpp"
#include "magic_enum/magic_enum.hpp"
#include "readCon/include/ReadCon.hpp"

namespace eonc::mat {

void from_con(Matter &mat, const std::string &fname) {
  std::vector<std::string> fconts =
      yodecon::helpers::file::read_con_file(fname);
  auto singleCon =
      yodecon::create_single_con<yodecon::types::ConFrameVec>(fconts);
  const size_t natoms = singleCon.atom_id.size();
  mat.nAtoms = natoms;
  mat.resize(natoms);

  // Set the cell and its inverse
  // TODO(rg) : Refactor out
  Matrix3S cell{Matrix3S::Zero()}, cellInverse{Matrix3S::Zero()};
  if (singleCon.angles[0] == 90.0 && singleCon.angles[1] == 90.0 &&
      singleCon.angles[2] == 90.0) {
    cell(0, 0) = singleCon.boxl[0];
    cell(1, 1) = singleCon.boxl[1];
    cell(2, 2) = singleCon.boxl[2];
  } else {
    singleCon.angles[0] *= M_PI / 180.0;
    singleCon.angles[1] *= M_PI / 180.0;
    singleCon.angles[2] *= M_PI / 180.0;

    cell(0, 0) = 1.0;
    cell(1, 0) = cos(singleCon.angles[0]);
    cell(1, 1) = sin(singleCon.angles[0]);
    cell(2, 0) = cos(singleCon.angles[1]);
    cell(2, 1) =
        (cos(singleCon.angles[2]) - cell(1, 0) * cell(2, 0)) / cell(1, 1);
    cell(2, 2) = sqrt(1.0 - pow(cell(2, 0), 2) - pow(cell(2, 1), 2));

    cell(0, 0) *= singleCon.boxl[0];
    cell(1, 0) *= singleCon.boxl[1];
    cell(1, 1) *= singleCon.boxl[1];
    cell(2, 0) *= singleCon.boxl[2];
    cell(2, 1) *= singleCon.boxl[2];
    cell(2, 2) *= singleCon.boxl[2];
  }
  cellInverse = cell.inverse();
  mat.cell = cell;
  mat.cellInverse = cellInverse;

  // Set other quantities
  AtomMatrix pos(natoms, 3);
  VectorType masses(natoms);
  Vector<size_t> atmnrs(natoms);
  Vector<int> isFixed(natoms);
  for (size_t idx{0}; idx < natoms; idx++) {
    // Needed since std::vector<bool> doesn't have .data()
    isFixed[idx] = static_cast<int>(singleCon.is_fixed[idx]);
    pos(idx, 0) = singleCon.x[idx];
    pos(idx, 1) = singleCon.y[idx];
    pos(idx, 2) = singleCon.z[idx];
    const auto elem =
        magic_enum::enum_cast<Element>(singleCon.symbol[idx]).value();
    // value() will fail anyway if an unsupported element is found
    const auto edat = elementData.find(elem)->second;
    masses[idx] = edat.atomicMass;
    atmnrs[idx] = edat.atomicNumber;
  }
  mat.setMasses(masses);
  mat.setAtomicNrs(atmnrs);
  mat.setFixedMask(isFixed);
  // Automatically applies periodic boundaries and sets recomputePotential
  mat.setPositions(pos);
}

} // namespace eonc::mat
