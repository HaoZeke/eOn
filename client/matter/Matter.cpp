/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#include "client/matter/Matter.h"
#include "client/thirdparty/xxhash.hpp"
#include <cachelot/hash_fnv1a.h>
#include <stdexcept>
// #include "BondBoost.h"

namespace eonc {

Matter::Matter(const Matter &matter) { operator=(matter); }

const Matter &Matter::operator=(const Matter &matter) {
  nAtoms = matter.nAtoms;
  resize(nAtoms);

  positions = matter.positions;
  forces = matter.forces;
  masses = matter.masses;
  atomicNrs = matter.atomicNrs;
  isFixed = matter.isFixed;
  cell = matter.cell;
  cellInverse = matter.cellInverse;
  velocities = matter.velocities;

  removeNetForce = matter.removeNetForce;
  usePeriodicBoundaries = matter.usePeriodicBoundaries;

  potential = matter.potential;
  potentialEnergy = matter.potentialEnergy;
  myCache = matter.myCache;

  return *this;
}

// The == comparison considers identity. This is crucial for process search.
// bool Matter::operator==(const Matter& matter) {
//     if(parameters->checkRotation) {
//         return helper_functions::rotationMatch(this, &matter,
//         parameters->distanceDifference);
//     }else{
//         return (parameters->distanceDifference) > perAtomNorm(matter);
//     }
// }

// bool Matter::operator!=(const Matter& matter) {
//     return !operator==(matter);
// }

// Returns the distance to the given matter object.
double Matter::distanceTo(const Matter &matter) {
  return pbc(positions - matter.positions).norm();
}

AtomMatrix Matter::pbc(AtomMatrix diff) const {
  AtomMatrix ddiff = diff * cellInverse;

  int i, j;
  for (i = 0; i < diff.rows(); i++) {
    for (j = 0; j < 3; j++) {
      ddiff(i, j) = fmod(fmod(ddiff(i, j), 1.0) + 1.5, 1.0) - .5;
    }
  }

  return ddiff * cell;
}

VectorType Matter::pbcV(VectorType diffVector) const {
  AtomMatrix pbcMatrix =
      pbc(AtomMatrix::Map(diffVector.data(), diffVector.size() / 3, 3));
  return VectorType::Map(pbcMatrix.data(), diffVector.size());
}

// Returns the maximum distance between two atoms in the Matter objects.
double Matter::perAtomNorm(const Matter &matter) const {
  size_t i{0};
  double max_distance = 0.0;

  if (matter.numberOfAtoms() == nAtoms) {
    AtomMatrix diff = pbc(positions - matter.positions);
    for (i = 0; i < nAtoms; i++) {
      max_distance = std::max(diff.row(i).norm(), max_distance);
    }
  }
  return max_distance;
}

void Matter::resize(const long int length) {
  if (length > 0) {
    nAtoms = length;
    positions.resize(length, 3);
    positions.setZero();

    velocities.resize(length, 3);
    velocities.setZero();

    biasForces.resize(length, 3);
    biasForces.setZero();

    forces.resize(length, 3);
    forces.setZero();

    masses.resize(length);
    masses.setZero();

    atomicNrs.resize(length);
    atomicNrs.setZero();

    isFixed.resize(length);
    isFixed.setZero();
  }
}

size_t Matter::numberOfAtoms() const { return (nAtoms); }

double Matter::getPosition(long int indexAtom, int axis) const {
  return positions(indexAtom, axis);
}

void Matter::setPosition(long int indexAtom, int axis, double position) {
  positions(indexAtom, axis) = position;
  if (usePeriodicBoundaries) {
    applyPeriodicBoundary();
  }
}

void Matter::setVelocity(long int indexAtom, int axis, double vel) {
  velocities(indexAtom, axis) = vel;
}

// return coordinates of atoms in array 'pos'
AtomMatrix Matter::getPositions() const { return positions; }

VectorType Matter::getPositionsV() const {
  return VectorType::Map(positions.data(), 3 * numberOfAtoms());
}

AtomMatrix Matter::getPositionsFree() const {
  AtomMatrix ret(numberOfFreeAtoms(), 3);
  size_t i{0}, j{0};
  for (i = 0; i < nAtoms; i++) {
    if (!isFixed(i)) {
      ret.row(j) = positions.row(i);
      j++;
    }
  }
  return ret;
}

Vector<size_t> Matter::getAtomicNrsFree() const {
  return this->atomicNrs.array() * getFreeV().cast<size_t>().array();
}

// bool Matter::relax(bool quiet, bool writeMovie, bool checkpoint,
//                    string prefixMovie, string prefixCheckpoint) {
//   auto objf =
//       std::make_shared<MatterObjectiveFunction>(Matter(*this), parameters);
//   auto optim =
//       helpers::create::mkOptim(objf, parameters->optim.method, parameters);

//   ostringstream min;
//   min << prefixMovie;
//   if (writeMovie) {
//     matter2con(min.str(), false);
//   }

//   int iteration = 0;
//   if (!quiet) {
//     SPDLOG_LOGGER_DEBUG(m_log, "{} {:10s}  {:14s}  {:18s}  {:13s}\n",
//                         "[Matter]", "Iter", "Step size",
//                         parameters->optim.convergenceMetricLabel, "Energy");
//     SPDLOG_LOGGER_DEBUG(m_log, "{} {:10}  {:14.5e}  {:18.5e}  {:13.5f}\n",
//                         "[Matter]", iteration, 0.0, objf->getConvergence(),
//                         getPotentialEnergy());
//   }

//   while (!objf->isConverged() && iteration < parameters->optim.maxIterations)
//   {

//     AtomMatrix pos = getPositions();

//     optim->step(parameters->optim.maxMove);
//     iteration++;
//     setPositionsFreeV(objf->getPositions());

//     double stepSize =
//         helper_functions::maxAtomMotion(pbc(getPositions() - pos));

//     if (!quiet) {
//       SPDLOG_LOGGER_DEBUG(m_log, "{} {:10}  {:14.5e}  {:18.5e}  {:13.5f}",
//                           "[Matter]", iteration, stepSize,
//                           objf->getConvergence(), getPotentialEnergy());
//     }

//     if (writeMovie) {
//       matter2con(min.str(), true);
//     }

//     if (checkpoint) {
//       ostringstream chk;
//       chk << prefixCheckpoint << "_cp";
//       matter2con(chk.str(), false);
//     }
//   }

//   if (iteration == 0) {
//     if (!quiet) {
//       SPDLOG_LOGGER_DEBUG(m_log, "{} {:10}  {:14.5e}  {:18.5e}  {:13.5f}",
//                           "[Matter]", iteration, 0.0, objf->getConvergence(),
//                           getPotentialEnergy());
//     }
//   }
//   //    bool converged = optimizer->run(parameters->optMaxIterations,
//   //    parameters->optMaxMove);
//   return objf->isConverged();
// }

VectorType Matter::getPositionsFreeV() const {
  return VectorType::Map(getPositionsFree().data(), 3 * numberOfFreeAtoms());
}

// update Matter with the new positions of the free atoms given in array 'pos'
void Matter::setPositions(const AtomMatrix pos) {
  positions = pos;
  if (usePeriodicBoundaries) {
    applyPeriodicBoundary();
  }
}

// Same but takes vector instead of n x 3 matrix
void Matter::setPositionsV(const VectorType pos) {
  setPositions(AtomMatrix::Map(pos.data(), numberOfAtoms(), 3));
}

void Matter::setPositionsFree(const AtomMatrix pos) {
  // FIXME: Ensure pos and existing data are in the same form with atom ids
  size_t i{0}, j{0};
  for (i = 0; i < nAtoms; i++) {
    if (!isFixed(i)) {
      positions.row(i) = pos.row(j);
      j += 1;
    }
  }
}

void Matter::setPositionsFreeV(const VectorType pos) {
  setPositionsFree(AtomMatrix::Map(pos.data(), numberOfFreeAtoms(), 3));
}

// AtomMatrix Matter::getBiasForces() {
//   if (biasPotential != NULL) {
//     biasPotential->boost();
//   }
//   return biasForces.array() * getFree().array();
// }

// void Matter::setBiasPotential(BondBoost *bondBoost) {
//   biasPotential = bondBoost;
// }

// void Matter::setBiasForces(const AtomMatrix bf) {
//   biasForces = bf.array() * getFree().array();
// }
// return forces applied on all atoms in array 'force'
AtomMatrix Matter::getForces() {
  computePotential();
  AtomMatrix ret = forces;
  size_t i{0};
  for (i = 0; i < nAtoms; i++) {
    if (isFixed[i]) {
      ret.row(i).setZero();
    }
  }
  return ret;
}

VectorType Matter::getForcesV() {
  return VectorType::Map(getForces().data(), 3 * numberOfAtoms());
}

AtomMatrix Matter::getForcesFree() {
  AtomMatrix allForces = getForces();
  AtomMatrix ret(numberOfFreeAtoms(), 3);
  size_t i{0}, j{0};
  for (i = 0; i < nAtoms; i++) {
    if (!isFixed[i]) {
      ret.row(j) = allForces.row(i);
      j++;
    }
  }
  return ret;
}

VectorType Matter::getForcesFreeV() {
  return VectorType::Map(getForcesFree().data(), 3 * numberOfFreeAtoms());
}

// return distance between the atoms with index1 and index2
double Matter::distance(long index1, long index2) const {
  return pbc(positions.row(index1) - positions.row(index2)).norm();
}

// return projected distance between the atoms with index1 and index2 on axis
// (0-x,1-y,2-z)
double Matter::pdistance(long index1, long index2, int axis) const {
  MatrixTRC<double, 1, 3> ret;
  ret.setZero();
  ret(0, axis) = positions(index1, axis) - positions(index2, axis);
  ret = pbc(ret);
  return ret(0, axis);
}

// return the distance atom with index has moved between the current Matter
// object and the Matter object passed as argument
double Matter::distance(const Matter &matter, long index) const {
  return pbc(positions.row(index) - matter.getPositions().row(index)).norm();
}

double Matter::getMass(long int indexAtom) const { return (masses[indexAtom]); }

void Matter::setMass(long int indexAtom, double mass) {
  masses[indexAtom] = mass;
}

void Matter::setMasses(VectorType massesIn) {
  for (size_t i = 0; i < nAtoms; i++) {
    masses[i] = massesIn[i];
  }
}

size_t Matter::getAtomicNr(size_t indexAtom) const {
  return (atomicNrs[indexAtom]);
}

void Matter::setAtomicNr(long int indexAtom, long atomicNr) {
  atomicNrs[indexAtom] = atomicNr;
}

int Matter::getFixed(long int indexAtom) const { return (isFixed[indexAtom]); }

void Matter::setFixed(long int indexAtom, int isFixed_passed) {
  isFixed[indexAtom] = isFixed_passed;
}

void Matter::setFixedMask(const Vector<int> &fixmask) { isFixed = fixmask; }
// void Matter::setPotentialEnergy(double epot_input)
//{
//	potentialEnergy=epot_input;
// }

double Matter::getPotentialEnergy() {
#ifdef EON_CHECKS
  if (nAtoms <= 0) {
    throw std::runtime_error("Called for a potential with 0 atoms");
  }
#endif
  computePotential();
  return potentialEnergy;
}

double Matter::getKineticEnergy() const {
  double K = 0;
  for (size_t i = 0; i < nAtoms; i++) {
    if (!isFixed[i])
      K += masses[i] * 0.5 * velocities.row(i).squaredNorm();
  }
  return K;
}

double Matter::getMechanicalEnergy() {
  return getPotentialEnergy() + getKineticEnergy();
}

long int Matter::numberOfFreeAtoms() const { return nAtoms - isFixed.sum(); }

long int Matter::numberOfFixedAtoms() const { return isFixed.sum(); }

size_t Matter::getForceCalls() const { return this->forceCalls; }

void Matter::resetForceCalls() {
  forceCalls = 0;
  return;
}

void Matter::computePotential() {
  size_t currentHash = computeHash();
  cachelot::cache::HashFunction calc_hash =
      cachelot::fnv1a<cachelot::cache::Cache::hash_type>::hasher();
  cachelot::slice cache_key(reinterpret_cast<const char *>(&currentHash),
                            sizeof(currentHash));

  // Try to retrieve from cache
  auto found_item = myCache->do_get(cache_key, calc_hash(cache_key));
  if (found_item) {
    // If found in cache, use the cached value
    std::tie(potentialEnergy, forces) =
        *(std::tuple<double, AtomMatrix> *)(found_item->value().str().c_str());
    std::cout << "Found " << potentialEnergy << std::endl;
    // SPDLOG_INFO("Found, so reusing {}",
    // std::string(found_item->value().begin(), found_item->value().end()));
  } else {
    // If not found in cache, compute potential energy
    auto tiedEF = potential->get_ef(positions, atomicNrs, cell);
    std::tie(potentialEnergy, forces) = tiedEF;
    // Store the computed value in the cache
    // TODO(rg): The size of this doesn't seem right, calculate the exact size
    // using Natoms and the rest.
    cachelot::slice value_slice((char *)(&tiedEF),
                                sizeof(std::tuple<double, AtomMatrix>));
    SPDLOG_INFO("Not found, so adding {}", potentialEnergy);
    auto new_item = myCache->create_item(cache_key, calc_hash(cache_key),
                                         value_slice.length(), 0,
                                         cachelot::cache::Item::infinite_TTL);
    new_item->assign_value(value_slice);
    myCache->do_set(new_item);
  }
}

// Transform the coordinate to use the minimum image convention.
void Matter::applyPeriodicBoundary() {
  AtomMatrix ddiff = positions * cellInverse;

  int i, j;
  for (i = 0; i < ddiff.rows(); i++) {
    for (j = 0; j < 3; j++) {
      ddiff(i, j) = fmod(ddiff(i, j) + 1.0, 1.0);
    }
  }
  positions = ddiff * cell;
}

double Matter::maxForce(void) {
  // Ensures that the forces are up to date
  computePotential();
  return (getForcesFree()).rowwise().norm().maxCoeff();
}

Vector<size_t> Matter::getAtomicNrs() const { return this->atomicNrs; }

void Matter::setAtomicNrs(const Vector<size_t> atmnrs) {
  if (static_cast<size_t>(atmnrs.size()) != this->nAtoms) {
    throw std::invalid_argument(
        "Vector of atomic numbers not equal to the number of atoms");
  } else {
    this->atomicNrs = atmnrs;
  }
}

AtomMatrix Matter::getFree() const {
  AtomMatrix ret(nAtoms, 3);
  size_t i{0}, j{0};
  for (i = 0; i < nAtoms; i++) {
    for (j = 0; j < 3; j++) {
      ret(i, j) = double(!bool(isFixed(i)));
    }
  }
  return ret;
}

VectorType Matter::getFreeV() const {
  return VectorType::Map(getFree().data(), 3 * numberOfAtoms());
}

AtomMatrix Matter::getVelocities() const {
  return velocities.array() * getFree().array();
}

void Matter::setVelocities(const AtomMatrix v) {
  velocities = v.array() * getFree().array();
}

void Matter::setForces(const AtomMatrix f) {
  forces = f.array() * getFree().array();
}

// AtomMatrix Matter::getAccelerations() {
//   AtomMatrix totF = getForces() + getBiasForces();
//   AtomMatrix ret = totF.array() * getFree().array();
//   ret.col(0).array() /= masses.array();
//   ret.col(1).array() /= masses.array();
//   ret.col(2).array() /= masses.array();
//   return ret;
// }

VectorType Matter::getMasses() const { return masses; }

void Matter::setPotential(std::shared_ptr<PotBase> pot) {
  this->potential = pot;
}

size_t Matter::getPotentialCalls() const { return this->npotcalls; }

double Matter::getEnergyVariance() { return this->energyVariance; }

// VectorType Matter::getForceVariance() {
//   return this->variance.segment(1, numberOfFreeAtoms() * 3);
// }

// double Matter::getMaxVariance() { return this->variance.maxCoeff(); }

std::shared_ptr<PotBase> Matter::getPotential() { return this->potential; }

size_t Matter::computeHash() const {
  // TODO(rg) :: Might get away with 32 bit
  xxh::hash_state_t<64> hash_stream;
  for (auto idx = 0; idx < positions.size(); ++idx) {
    hash_stream.update(reinterpret_cast<const char *>(&positions.data()[idx]),
                       sizeof(positions.data()[idx]));
  }
  for (auto idx = 0; idx < isFixed.size(); ++idx) {
    hash_stream.update(reinterpret_cast<const char *>(&isFixed[idx]),
                       sizeof(isFixed[idx]));
  }
  return hash_stream.digest();
}

} // namespace eonc
