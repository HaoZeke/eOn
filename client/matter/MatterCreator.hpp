/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#pragma once
#include "client/matter/Matter.h"

namespace eonc::mat {

void from_con(Matter &mat, const std::string &fname);

} // namespace eonc::mat
