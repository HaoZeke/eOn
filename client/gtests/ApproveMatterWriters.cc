/*
** This file is part of eOn.
**
** SPDX-License-Identifier: BSD-3-Clause
**
** Copyright (c) 2010--present, eOn Development Team
** All rights reserved.
**
** Repo:
** https://github.com/TheochemUI/eOn
*/
#include "TestApprovalMain.hpp"

#include "client/io/WriteCreator.hpp"
#include "client/matter/Matter.h"
#include "client/matter/MatterCreator.hpp"

#include <iostream>
#include <string>

std::string readFileContent(const std::string &filename) {
  std::ifstream file(filename);
  std::stringstream buffer;
  buffer << file.rdbuf();
  return buffer.str();
}

std::vector<eonc::Matter> getTestMatter() {
  const auto config =
      toml::table{{"Potential", toml::table{{"potential", "lj"}}}};
  auto pot_default = eonc::makePotential(config);
  auto cachelot = cachelot::cache::Cache::Create(
      eonc::cache_memory, eonc::page_size, eonc::hash_initial, true);
  auto matter = eonc::Matter(pot_default, &cachelot);
  std::string confile("pos.con");
  eonc::mat::ConFileParser cfp;
  cfp.parse(matter, confile);
  return {matter};
}

TEST_CASE("VerifyMatter2Con") {
  auto testMatter = getTestMatter()[0];
  const auto config = toml::table{{"Main", toml::table{{"write", "con"}}}};
  std::string filename = "test_output.con";
  auto conWriter = eonc::io::mkWriter(config);
  conWriter->write(testMatter, filename);
  std::string fileContent = readFileContent(filename);
  ApprovalTests::Approvals::verify(fileContent);
}

TEST_CASE("VerifyMatter2XYZ") {
  auto testMatter = getTestMatter()[0];
  const auto config = toml::table{{"Main", toml::table{{"write", "xyz"}}}};
  std::string filename = "test_output.xyz";
  auto XYZWriter = eonc::io::mkWriter(config);
  XYZWriter->write(testMatter, filename);
  std::string fileContent = readFileContent(filename);
  ApprovalTests::Approvals::verify(fileContent);
}

TEST_CASE("VerifyMatter2Convel") {
  auto testMatter = getTestMatter()[0];
  const auto config = toml::table{{"Main", toml::table{{"write", "convel"}}}};
  testMatter.setVelocities(testMatter.getPositionsFree().array() + 3);
  std::string filename = "test_output.con";
  auto ConvelWriter = eonc::io::mkWriter(config);
  ConvelWriter->write(testMatter, filename);
  std::string fileContent = readFileContent(filename);
  ApprovalTests::Approvals::verify(fileContent);
}

TEST_CASE("VerifyMatter2Tibble") {
  auto testMatter = getTestMatter()[0];
  const auto config = toml::table{{"Main", toml::table{{"write", "tibble"}}}};
  std::string filename = "test_output.txt";
  auto TibbleWriter = eonc::io::mkWriter(config);
  TibbleWriter->write(testMatter, filename);
  std::string fileContent = readFileContent(filename);
  ApprovalTests::Approvals::verify(fileContent);
}
