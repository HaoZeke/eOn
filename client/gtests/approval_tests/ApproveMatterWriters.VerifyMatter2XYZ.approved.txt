13
Generated by eOn
O 10.477713 12.379463 12.871778
O 9.838316 12.065258 10.599445
C 13.632107 12.976546 11.590461
C 12.574398 13.442847 10.828786
C 13.839846 11.585323 11.720681
C 13.015406 10.695876 11.089027
H 14.180164 13.651700 12.258033
H 12.293139 14.494230 10.914210
H 12.410976 13.061416 9.811738
H 12.993463 9.652122 11.395309
H 14.546376 11.239529 12.484877
H 12.591733 10.859365 10.096371
S 10.972953 12.113958 11.523496
